Trigger an event when a repeatable-subform is sorted

ryandemmer opened the pull request:
Pull Request for Issue #24343 (Part 1 of 2)

### Summary of Changes
This PR adds a configuration option to the sortable function to trigger a "subform-row-sort" event when a repeatable subform is sorted.

### Testing Instructions
After applying the PR, make a copy of the file. Delete the media/system/js/subform-repeatable.js, and rename the copy to media/system/js/subform-repeatable.js

On the bottom of the file, under
```javascript
$(document).on('subform-row-add', initSubform);
```
add

```javascript
$(document).on('subform-row-sort', function (event, row) {
    console.log(row);
});
```

In Content -> Fields, create a new Field with the **Type** _Repeatable_. Give the Field a name, eg: _Repeat Text_. Create a **Form Field** of type _Text_ and give it a name.

Assign the field to **Category** _All_ and **Field Group** _None_.

Click **Save & Close**.

Create a new article to edit, click on the Fields tab. Click on the plus icon twice in Text Repeat to create 2 new Text fields.

Open your browsers javascript console.

Drag the bottom field using the blue handle to re-order it above the top field. When this is done you should see a message in the browser's Console.

After conducting the test, remove the logging script, ie: 

```javascript
$(document).on('subform-row-sort', function (event, row) {
    console.log(row);
});
```

### Actual result BEFORE applying this Pull Request
No message is disaplyed in the browser console as not sort event is triggered.


### Expected result AFTER applying this Pull Request
A sort event is triggered and a message is displayed in the browser console.


### Documentation Changes Required
None.

ryandemmer committed:
Trigger an event when a subform is sorted

pixxelfactory reviewed (changes requested):
I noticed that i had to switch the "update" event to "stop", so that it fires if the user drags the row without changing its order too, otherwise if the user drops the row at the same place, the event is not fired.

ryandemmer committed:
Update subform-repeatable-uncompressed.js

Change "update" sortable option to "stop" so event is triggered even if the sort order is not changed.

ryandemmer reviewed (commented):
Changed "update" option to "stop"

pixxelfactory commented:
This pull request is open for more than a year (and so is the bug in Joomla), can someone please merge it?

brianteeman commented:
every pull request needs at least two successful test before merging. This doesnt have any yet.

Fedik commented:
Additionally it should use "update" instead of "stop".
The event should not be triggered if user does not change the order.

ryandemmer commented:
> Additionally it should use "update" instead of "stop".
> The event should not be triggered if user does not change the order.

If you don't change the order but still attempt the drag, this can still affect the dom and therefore the editor, which is the point of triggering the event. See the reference to the issue in the initial comment.

> Btw, it is "new feature".

This is part of a fix for an issue, so is not a new feature.

Fedik commented:
There no point to trigger this event if the order does not changes.

pixxelfactory commented:
> 
> 
> There no point to trigger this event if the order does not changes.

Yes there is:
If someone drag and drops the item without changing the order, the content gets lost anyway if the event does not get triggered.
What @ryandemmer did here is completely fine and correct, trust me, i'm implementing his fix manually for over a year now on my sites.

Fedik commented:
Please prove, show a real case.

> If someone drag and drops the item without changing the order, the content gets lost anyway if the event does not get triggered.

What and where is lost?

ryandemmer commented:
@Fedik - Did you read the issue this refers to? - https://github.com/joomla/joomla-cms/issues/24343 - specifically - https://github.com/joomla/joomla-cms/issues/24343#issuecomment-515784534

Fedik commented:
I have checked, but this PR does not fix that issue.

My comment still valid:

> There no point to trigger this event if the order does not changes.

Fedik commented:
I will give you a hint, every sorting library have at least 3 events: 
start (always), change (only if something changed), stop (always)

If you want 1, then you required to have another 2.

Quy commented:
@Fedik Is this still a valid PR?

Fedik commented:
nope

Fedik commented:
I am closing it.
In current implementation it is not much useful.
If need "sorting" events, then should be all: start, stop, cancel (kind of)

Feel free to make new PR, for joomla 4.
